metrics.go

```diff
@@ -0,0 +1,207 @@
+}
+
+/*
+Return a client for a metric destination struct
+*/
+func clientConfigFromDestination(destination *metricsDestination) *clientConfig {
+	return &clientConfig{
+		Host:   destination.Host + ":" + destination.Port,
+		Prefix: destination.Prefix,
+	}
+}
+
+/*
+Struct responsible for building clients
+*/
+type clientBuilder struct {
+	errorLogger func(error)
+}
+
+/*
+Return a statsd client from our custom clientConfig
+statsd.new will resolve the given host and panics if it fails
+The statsd lib we use sends a message upon connection
+if there is nothing listening it may crash
+TODO : check that this code is removed https://github.com/alexcesaro/statsd/blob/master/conn.go#L50
+*/
+func (builder *clientBuilder) buildClient(config *clientConfig) *statsd.Client {
+	client, err := statsd.New(
+		statsd.Address(config.Host),
+		statsd.Prefix(config.Prefix),
+		statsd.ErrorHandler(builder.errorLogger),
+	)
+
+	if err != nil {
+		logger.WithFields(logrus.Fields{
+			"config": config,
+			"error":  err,
+		}).Error("Error creating the statsd client")
+		panic("Error creating statsd client")
+	}
+
+	return client
+}
+
+/*
+Initialize the client from the standard configuration of metrics (METRICS_HOST, METRICS_PORT, METRICS_PREFIX)
+*/
+func (metricsSender *metricsSender) initStandardClient(clientBuilder *clientBuilder) {
+	host := os.Getenv("METRICS_HOST")
+
+	if host != "" {
+		port := os.Getenv("METRICS_PORT")
+		prefix := os.Getenv("METRICS_PREFIX")
+
+		if port == "" || prefix == "" {
+			panic("Basic configuration can not have any empty port or prefix")
+		}
+
+		basicConfig := &clientConfig{
+			Host:   host + ":" + port,
+			Prefix: prefix,
+		}
+		metricsSender.Clients = append(metricsSender.Clients, clientBuilder.buildClient(basicConfig))
+	}
+}
+
+/*
+Initialize one or several clients from the METRICS_DESTINATIONS env var
+*/
+func (metricsSender *metricsSender) initDestinationClients(clientBuilder *clientBuilder) {
+	metricsDestinationVar := os.Getenv("METRICS_DESTINATIONS")
+
+	if metricsDestinationVar != "" {
+		destinations := &metricsDestinations{}
+		err := json.Unmarshal([]byte(metricsDestinationVar), destinations)
+
+		if err != nil {
+			logger.WithField("metrics_destinations", metricsDestinationVar).Error("Error parsing env METRICS_DESTINATION")
+			panic("Error creating statsd client - JSON unmarshalling")
+		}
+
+		for _, destination := range *destinations {
+			config := clientConfigFromDestination(destination)
+			metricsSender.Clients = append(metricsSender.Clients, clientBuilder.buildClient(config))
+		}
+	}
+}
+
+/*
+Initialize the clients from the two possible configurations
+*/
+func (metricsSender *metricsSender) initMetricsSender() {
+	metricsSender.initStandardClient(defaultClientBuilder)
+	metricsSender.initDestinationClients(defaultClientBuilder)
+}
+
+func GetMetricsSender() *metricsSender {
+	if len(metrics.Clients) == 0 {
+		metrics.initMetricsSender()
+	}
+	return metrics
+}
+
+func resetMetricsSender() {
```

Comment:
You should put that function in the test file directly, like that it will not be compiled in the production binary